python/bap.py

```diff
@@ -0,0 +1,342 @@
+
+class Id(object):
+    def __init__(self, r):
+        self.value = r
+    def __str__(self):
+        return str(self.value)
+
+RETRIES = 10
+
+class Bap(object):
+    def __init__(self, server={}):
+        if isinstance(server, dict):
+            self.__dict__.update(spawn_server(**server))
+        else:
+            self.url = server
+
+        self.last_id = 0
+        for attempt in range(RETRIES):
+            try:
+                self.capabilities = self.call({'init' : {
+                    'version' : '0.1'}})
+            except Exception:
+                if attempt + 1 == RETRIES:
+                    raise
+                else:
+                    time.sleep(0.1 * attempt)
+
+        if not "capabilities" in self.__dict__:
+            raise RuntimeError("Failed to connect to BAP server")
+
+        self.data = {}
+        self.temp = NamedTemporaryFile('rw+b')
+
+    def insns(self, src):
+        res = self.call({'get_insns' : {'resource' : src}})
+        for msg in res:
+            if 'error' in msg:
+                err = Error(msg)
+                if err.severity in DEBUG_LEVEL:
+                    print err
+            else:
+                return (parse_insn(js) for js in msg['insns'])
+
+    def close(self):
+        self.__exit__()
+
+    def load_file(self, name):
+        return self._load_resource({'load_file' : {
+            'url' : 'file://' + name}})
+
+    def get_resource(self, name):
+        return self.call({'get_resource' : name}).next()
+
+    def load_chunk(self, data, **kwargs):
+        kwargs.setdefault('url', self.mmap(data))
+        kwargs.setdefault('arch', 'x86_32')
+        kwargs.setdefault('address', bil.Int(0,32))
+        kwargs.setdefault('endian', bil.LittleEndian())
+        return self._load_resource({'load_memory_chunk' : kwargs})
+
+    def __exit__(self):
+        if 'server' in self.__dict__:
+            self.server.terminate()
+        self.temp.close()
+
+    def call(self, data):
+        def dumps(dic):
+            self.last_id += 1
+            dic['id'] = Id(self.last_id)
+            return json.dumps(dic, default=str)
+
+        if isinstance(data, dict):
+            method = requests.post
+            if 'get_insns' or 'get_resource' in data:
+                method = requests.get
+            return jsons(method(self.url, data=dumps(data)))
+        else:
+            gen = (dumps(msg) for msg in data)
+            return jsons(requests.post(self.uri, data=gen))
+
+    def mmap(self, data):
+        url = "mmap://{0}?offset=0&length={1}".format(
+            self.temp.name, len(data))
+        os.ftruncate(self.temp.fileno(), len(data))
+        mm = mmap(self.temp.fileno(), len(data))
+        mm.write(data)
+        mm.close()
+        return url
+
+    def _load_resource(self, res):
+        rep = self.call(res).next()
+        if 'error' in rep:
+            raise ServerError(rep)
+        return Id(rep['resource'])
+
+def spawn_server(**kwargs):
```

Comment:
`HTTP` is used for control information. `shm` is used to share binary data. Currently bap supports only `HTTP` as a control information transport. It can be easily extended, but I do not see any reason to pass requests and response via `shm` or some other faster mechanisms. Actually, I've traces, most of the time we spend in python, trying to understand bap's answer, not in sending or receiving data.